### ISSUE TITLE
SwarmSpawner,  InvalidArgument: Incompatible options have been provided for the bind type mount.

### ISSUE THREAD

**Issue by camm89** [jupyterhub/dockerspawner#263]
Trying to deploy hub in Docker Swarm on Docker UCP. We are able to to get the main hub up and running, but when we try to spawn the single user notebooks we get the following:

```
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |     Traceback (most recent call last):
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/tornado/web.py", line 1592, in _execute
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         result = yield result
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 1052, in get
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         await self.spawn_single_user(user)
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 705, in spawn_single_user
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         timedelta(seconds=self.slow_spawn_timeout), finish_spawn_future
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 626, in finish_user_spawn
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         await spawn_future
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 489, in spawn
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         raise e
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 409, in spawn
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/dockerspawner/dockerspawner.py", line 691, in start
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         obj = yield self.create_object()
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 173, in create_object
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         mounts=self.mounts,
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 110, in mounts
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         for host_loc, vol in self.volume_binds.items()
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 110, in <listcomp>
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         for host_loc, vol in self.volume_binds.items()
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |       File "/opt/conda/lib/python3.6/site-packages/docker/types/services.py", line 237, in __init__
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |         'Incompatible options have been provided for the bind '
EAE_jupyterhub.1.zluhmt93kq7u@linuxkit-025000000001    |     docker.errors.InvalidArgument: Incompatible options have been provided for the bind type mount.
```

jupyterhub-config had the following lines to handle volumes

```python
notebook_dir = '/home/jovyan'
c.SwarmSpawner.notebook_dir = notebook_dir
c.SwarmSpawner.volumes = {'/mnt/our_shared_path': notebook_dir, 'mode': 'rw'}
c.SwarmSpawner.volume_driver = None
```

Tried commenting out the 'SwarmSpawner.volumes' line but we still see the same error as above. 
According to binds in swarmspawner.py (https://github.com/jupyterhub/dockerspawner/blob/master/dockerspawner/swarmspawner.py#L99), if len(self.volume_binds) == 0 it should just return [] and there should be no error. We do not create a mount in docker-compose or manually on the host. Using Jupyterhub version 0.9.4. Haven't been able to find a way around this issue.

**Comment by wlohman-surfsara** [comment]
I get the exact same

Problem seems to be here: /opt/conda/lib/python3.6/site-packages/docker/types/services.py

When I comment out the if statement starting at line 237, it works without issue:

            #if any([labels, driver_config, no_copy, tmpfs_size, tmpfs_mode]):
            #    raise errors.InvalidArgument(
            #        'Incompatible options have been provided for the bind '
            #        'type mount.'
            #    )

Adding some debugging statements there:

            print(f"labels: {labels}")
            print(f"driver_config: {driver_config}")
            print(f"no_copy: {no_copy}")
            print(f"tmpfs_size: {tmpfs_size}")
            print(f"tmpfs_mode: {tmpfs_mode}")

Gives:

        labels: None
        driver_config: {'Name': ''}
        no_copy: False
        tmpfs_size: None
        tmpfs_mode: None

So apparently it is the driver_config containing anything, even an empty Name, that is causing the problem, so I should be able to do this:

            #if any([labels, driver_config, no_copy, tmpfs_size, tmpfs_mode]):
            if any([labels, no_copy, tmpfs_size, tmpfs_mode]):

Now it also works. The container is spawned and the mount exists.

Clearly when the type == 'bind' services.py does not accept any value in driver_config, and throws an error if it does. And as 'bind' is hardcoded here there seems little use setting driver_config.

So, I revert all my changes and change line 108 in swarmspawner.py:

            return [
                Mount(
                    target=vol["bind"],
                    source=host_loc,
                    type="bind",
                    read_only=vol["mode"] == "ro",
                     #driver_config=driver,
                     driver_config=None,
                )

With this I also get my container to spawn, with my mounts. It's not a nice solution, and not one I think is intended by the author.

**Comment by camm89** [comment]
@smparekh

**Comment by dprittie** [comment]
I think I am running into the same problem

I would like to use a volume driver other than local but with the following config I get the same error
```
notebook_dir = '/home/jovyan'
c.SwarmSpawner.notebook_dir = notebook_dir
c.SwarmSpawner.volumes = { 'jupyterhub-user-{username}': notebook_dir }
c.SwarmSpawner.volume_driver = 'nimble'```

```
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/tornado/web.py", line 1543, in _execute
        result = yield result
      File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 1012, in get
        await self.spawn_single_user(user)
      File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 671, in spawn_single_user
        await gen.with_timeout(timedelta(seconds=self.slow_spawn_timeout), finish_spawn_future)
      File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 626, in finish_user_spawn
        await spawn_future
      File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 486, in spawn
        raise e
      File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 406, in spawn
        url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
      File "/opt/conda/lib/python3.6/site-packages/dockerspawner/dockerspawner.py", line 691, in start
        obj = yield self.create_object()
      File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 173, in create_object
        mounts=self.mounts,
      File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 110, in mounts
        for host_loc, vol in self.volume_binds.items()
      File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 110, in <listcomp>
        for host_loc, vol in self.volume_binds.items()
      File "/opt/conda/lib/python3.6/site-packages/docker/types/services.py", line 237, in __init__
        'Incompatible options have been provided for the bind '
    docker.errors.InvalidArgument: Incompatible options have been provided for the bind type mount.
```

**Comment by vlizanae** [comment]
@wlohman-surfsara I tried this by overwriting the `mounts` in the config file:

```
from dockerspawner import SwarmSpawner
from docker.types import Mount

class SwarmSpawner2(SwarmSpawner):
    @property
    def mounts(self):
        if len(self.volume_binds):
            driver = self.mount_driver_config
            return [
                Mount(
                    target=vol["bind"],
                    source=host_loc,
                    type="bind",
                    read_only=vol["mode"] == "ro",
                    driver_config=None,
                )
                for host_loc, vol in self.volume_binds.items()
            ]

        else:
            return []

c.JupyterHub.spawner_class = SwarmSpawner2

c.SwarmSpawner2.network_name = os.environ['DOCKER_NETWORK_NAME']
c.SwarmSpawner2.extra_host_config = {'network_mode': os.environ['DOCKER_NETWORK_NAME']}

c.SwarmSpawner2.image = 'registry:5000/jupyterhub_singleuser'

c.SwarmSpawner2.notebook_dir = '/home/jovyan'
c.SwarmSpawner2.volumes = {'/data/home/{username}': '/home/jovyan'}
c.SwarmSpawner2.remove_containers = True
```
When I comment the line with `c.SwarmSpawner2.volumes` works (without persistency obviously) but when it is like all I get is
```
Spawner failed to start [status=0]
```

**Comment by vlizanae** [comment]
It works now, it was a permissions issue, but this is definitely a bug, should be labeled as such.

**Comment by hugoJuhel** [comment]
Hi @vlizanae , i'm facing the same issue, while trying to mount a NFS directory (directly mounted in jupyterhub using Netshare plugin). I have used your `SwarmSpawner2` class, but there is no logs in the spawned service. Could you please telle me which was the permission issues please ? Thanks a lot .

**Comment by vlizanae** [comment]
The users' folders must have owner `1000:100` at least in my case, this is because the image I'm spawning comes from the jupyter docker-stacks, which always use jovyan as user.

**Comment by hugoJuhel** [comment]
Hi @vlizanae 

Thanks for your answer. I did as you mentionned but it still not working. I'm on ubuntu so I also used `chown 1000:100`.  I'm also using a jupyter-stack notebook.

The permissions are : 
- In juperhub : (I have mounted the NFS in /mnt/{username}):
`drwxrwxrwx   3 1000 users 4096 Mar  5 21:41 mnt`

- On the NFS server, 
`drwxrwxrwx 3 1000 users 4096 Mar  5 10:07 jupyterUser`

But i still have the same error. My NFS is correctly mounted on the NFS. 

I'm missing something but I dont't see it. Do you have any idea ? 

Thanks !

**Comment by vlizanae** [comment]
You don't have to mount the volumes in the jupyterhub service, the jupyterhub service uses docker to mount the volumes from the host, in the jupyterhub config you have to put the path to the volumes as seen from the host.

If that is already the case I don't know what could be, what I did to find the error was to put the jupyterhub in an attachable docker network and try to spawn myself a container attached to that same network, with docker's `inspect` you can see the parameters needed, like the jupyterhub token for example.

**Comment by hugoJuhel** [comment]
@vlizanae  my bad. Thanks a lot, this is working know. I mounted on each node the NFS file, and I now adequatly bind  the host dir to the jupyter notebook one.

**Comment by zhiyuli** [comment]
@camm89 
You found a fix? I tried all the solutions above (change services.py or subclass SwarmSpawner or change folder to 1000:100), but neither worked. Thanks

**Comment by smparekh** [comment]
@zhiyuli ,

We subclassed `SwarmSpawner` and override the `mounts` method:
```python
    @property
    def mounts(self):
        if len(self.volume_binds):
            driver = self.mount_driver_config
            return [
                Mount(
                    target=vol["bind"],
                    source=host_loc,
                    type="bind",
                    read_only=False,
                    driver_config=None,
                )
                for host_loc, vol in self.volume_binds.items()
            ]

        else:
            return []

```

**Comment by zhiyuli** [comment]
@smparekh 
Thanks very much for the quick follow-up and confirming the subclass worked for you. I double checked my codes and found I forgot to "import Mount"! 
It is working like a charm now! Thanks!

**Comment by minrk** [comment]
Hi! I’m going through and cleaning up old/stale issues on this repo. I believe this should be fixed by #406 which adds support for the mounts spec.

**Comment by cmotadev** [comment]
I solved this issue:

1) Creation of a module "modspawner.py"

`
from dockerspawner import SwarmSpawner

class ModSwarmSpawner(SwarmSpawner):
    @property
    def mount_driver_config(self):
        if self.volume_driver:
            return DriverConfig(
                name=self.volume_driver, options=self.volume_driver_options or None
            ) 
        return None
`

2) Add to JupyterHub settings:

`
c.JupyterHub.spawner_class = 'modspawner.ModSwarmSpawner'
`

The main reason is because of a improprely implementation of mount_driver_config method, of the SwarmSpawner Class.
This methods always returns a DriverConfig, even if no driver config was set. 

When the spawner calls to create a Service, it creates a docker.types.services.Mount Object, and it checks if driver_config evaluates False. Due to the driver_config always returns a DriverConfig Object (that evaluates True), it always throws a "'Incompatible options have been provided for the bind type mount." error.  This subclass solves this problem and I will make a pull request to aplly this on the main repo.

References: 
https://ask.csdn.net/questions/5861756
https://github.com/docker/docker-py/blob/31775a1532a66cf8a4c183a99bb5c73623147295/docker/types/services.py#L240
